### PR TITLE
groups: REST API and API addition

### PR DESCRIPTION
**Pull request by hachreak**
* Adds API and REST API for the groups.

* Implements a policy for "pending" and "member" state of user
  in a group.

* Removes current_user dependency in accounts model.

* Refactors the views to use api instead of access directly to
  the models.

Signed-off-by: Konstantinos Kostis <konstantinos.kostis@cern.ch>
Reviewed-by: Leonardo Rossi <leonardo.r@cern.ch>

### CODE REVIEW

File: invenio/modules/accounts/models.py [hunk 1]
```diff
@@ -184,65 +211,95 @@ def login_method_is_external(self):
...
+    @session_manager
+    def approve(self, user):
+        """Approve user from group.
+
+        :param user: User to approve into the group.
+        """
+        # approve user
+        uug = UserUsergroup.query.filter(
+            UserUsergroup.id_user.like(user.id),
+            UserUsergroup.id_usergroup.like(self.id),
```
Comment (jirikuncar) [inline invenio/modules/accounts/models.py]: maybe you could do it at once with update:
```python
UserUsergroup.query.filter(
    UserUsergroup.id_user.like(user.id),
    UserUsergroup.id_usergroup.like(self.id),
    UserUsergroup.user_status == UserUsergroup.USER_STATUS['PENDING']
).update(dict(user_status=UserUsergroup.USER_STATUS['MEMBER'])
```

File: invenio/modules/accounts/models.py [hunk 2]
```diff
@@ -269,10 +326,66 @@ def get_users_not_in_this_group(self, nickname=None, email=None,
...
+        except NoResultFound:
+            raise IntegrityUsergroupError(
+                'User with id "{0}" not exists').format(id_user_admin)
+        group.join_policy = group.join_policy or \
+            Usergroup.JOIN_POLICIES['VISIBLEOPEN']
+        group.login_method = group.login_method or \
+            Usergroup.LOGIN_METHODS['INTERNAL']
+        group.description = group.description or ''
+        # create the new group
+        db.session.add(group)
```
Comment (jirikuncar) [inline invenio/modules/accounts/models.py]: IMHO you should join the user to the group before commit

Comment (hachreak) [inline invenio/modules/accounts/models.py]: done.

File: invenio/modules/groups/api.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
+# -*- coding: utf-8 -*-
```
Comment (lnielsen) [inline invenio/modules/groups/api.py]: Couldn't a lot of these methods be class methods on the individual model classes? It makes it much more clear what you are dealing with.

**Comment by lnielsen** [comment]
We can integrate this PR, but I'd like to have more general discussion regarding login method/join policy plus class vs function based apis, hence current PR should not set the API in stone :-) 

cc @bouzlibop

**Comment by jirikuncar** [comment]
> class vs function based apis

I would say we should stick with something similar to `invenio.modules.records.api:Record` - hence I'm in favor of class based APIs.

File: invenio/modules/groups/api.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    :param id_user: user's id
+    :param group_name: string to filter group list
+    :return filtering query
+    """
+    return models_get_groups_user_not_joined(
+        id_user=id_user,
+        group_name="%%%s%%" % group_name)
+
+
+def query_members_of_group(id_usergroup, curr_uid=None):
```
Comment (jirikuncar) [inline invenio/modules/groups/api.py]: `curr_uid` should be `id_user`

**Comment by hachreak** [comment]
@jirikuncar  I refactored the `api.py` code (and adapted `restul.py` and `views.py`)

**Comment by lnielsen** [comment]
This PR is superseded by inveniosoftware/invenio-groups#1

**Pull request closed without merging**